Add Gen 9 BSS Factory

LegoFigure11 opened the pull request:
This pull request adds Gen 9 BSS Factory.

Sets by @NO-VED and the rest of the BSS Factory team. Full credits can be found in the forum thread that will be made public once the format goes live.

Apologies if there are any formatting or code style issues, I couldn't get eslint to play nice on my machine. Passes all tests but if there are other issues please let me know and I'll try to amend them in the coming days!

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+          ["protect"]
+        ],
+        "item": ["toxicorb"],
+        "nature": "impish",
+        "evs": [204, 4, 252, 0, 28, 20],
```
-This breloom has no attacks, and therefore does not need attack EVs, IVs, or nature. Amend this.
-20 speed doesn't do anything, Honchkrow and Sandaconda don't exist. 204/0/252/0/48/4 would be acceptable.

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        "moves": [["curse"], ["rest"], ["wavecrash"], ["fissure"]],
+        "item": ["chestoberry"],
+        "nature": "careful",
+        "evs": [252, 0, 0, 0, 252, 4],
```
Please don't put the 4 in Speed on max/max sets in randoms. Move it elsewhere.

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        ],
+        "item": ["rockyhelmet", "leftovers"],
+        "nature": "impish",
+        "evs": [252, 0, 252, 0, 0, 4],
```
Please don't put the 4 in Speed on max/max sets in randoms. Move it elsewhere.

Also applies to other corv sets.

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        ],
+        "item": ["assaultvest"],
+        "nature": "quiet",
+        "evs": [252, 0, 0, 252, 0, 4],
```
Please don't put the 4 in Speed on max/max sets in randoms. Move it elsewhere.

Also applies to other Hoodras.

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        "moves": [["earthquake"], ["avalanche"], ["yawn"], ["protect"]],
+        "item": ["leftovers"],
+        "nature": "impish",
+        "evs": [156, 4, 252, 0, 92, 4],
```
Probably just move the attack and speed into HP or SpD.

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        "moves": [["aquajet"], ["playrough"], ["liquidation"], ["superpower"]],
+        "item": ["assaultvest"],
+        "nature": "adamant",
+        "evs": [252, 252, 0, 0, 0, 4],
```
Please don't put the 4 in Speed on max/max sets in randoms. Move it elsewhere. Also applies to other Azumarill sets. Move the 12s, too.

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        ],
+        "item": ["assaultvest"],
+        "nature": "adamant",
+        "evs": [252, 252, 0, 0, 0, 4],
```
Please don't put the 4 in Speed on max/max sets in randoms. Move it elsewhere. Also applies for other Kingambits.

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        "ability": "defiant"
+      },
+      {
+        "species": "kingambit",
+        "weight": 5,
+        "moves": [
+          ["stealthrock"],
+          ["kowtowcleave"],
+          ["ironhead", "guillotine"],
+          ["suckerpunch", "thunderwave"]
+        ],
+        "item": ["sitrusberry", "focussash"],
+        "nature": "adamant",
+        "evs": [244, 252, 4, 0, 4, 4],
+        "teratype": ["flying", "fairy", "dark"],
+        "ability": "defiant"
+      },
+      {
+        "species": "kingambit",
+        "weight": 5,
+        "moves": [
+          ["stealthrock"],
+          ["kowtowcleave"],
+          ["ironhead", "guillotine"],
+          ["suckerpunch", "thunderwave"]
+        ],
+        "item": ["sitrusberry", "focussash"],
+        "nature": "adamant",
+        "evs": [244, 252, 4, 0, 4, 4],
+        "teratype": ["flying", "fairy", "dark"],
+        "ability": "supremeoverlord"
```
```suggestion
        "species": "kingambit",
        "weight": 5,
        "moves": [
          ["stealthrock"],
          ["kowtowcleave"],
          ["ironhead", "guillotine"],
          ["suckerpunch", "thunderwave"]
        ],
        "item": "focussash",
        "nature": "adamant",
        "evs": [252, 252, 0, 0, 4, 0],
        "teratype": ["flying", "fairy", "dark"],
        "ability": ["defiant", "supremeoverlord"]
      },
      {
        "species": "kingambit",
        "weight": 5,
        "moves": [
          ["stealthrock"],
          ["kowtowcleave"],
          ["ironhead", "guillotine"],
          ["suckerpunch", "thunderwave"]
        ],
        "item": "sitrusberry",
        "nature": "adamant",
        "evs": [244, 252, 0, 0, 12, 0],
        "teratype": ["flying", "fairy", "dark"],
        "ability": ["defiant", "supremeoverlord"]
```

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        ],
+        "item": ["leftovers"],
+        "nature": "bold",
+        "evs": [252, 0, 252, 0, 0, 4],
```
Please don't put the 4 in Speed on max/max sets in randoms. Move it elsewhere. Also applies to other Wo-Chiens.

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        ],
+        "item": ["leftovers", "punchingglove"],
+        "nature": "impish",
+        "evs": [92, 4, 172, 0, 212, 28],
```
Only 20 EVs are required to outspeed Bloodmoon.

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        ],
+        "item": ["sitrusberry", "rockyhelmet"],
+        "nature": "impish",
+        "evs": [252, 0, 252, 0, 0, 4],
```
move the speed elsewhere on all orthworm

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        ],
+        "item": ["loadeddice"],
+        "nature": "adamant",
+        "evs": [252, 252, 0, 0, 0, 4],
```
Move the 4 EVs elsewhere.

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+          ["knockoff"]
+        ],
+        "item": ["assaultvest"],
+        "nature": "jolly",
+        "evs": [252, 4, 4, 0, 244, 4],
```
Jolly Breloom does not exist in this format. This spread is useless unless you have Rapid Spin, which is a 1/3 chance. Just run max/max Careful with 4 attack or something. A different spread.

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        ],
+        "item": ["leftovers"],
+        "nature": "careful",
+        "evs": [244, 4, 4, 0, 252, 4],
```
```suggestion
        "evs": [252, 4, 0, 0, 252, 0],
```
Lefties numbers aren't a thing.

ACakeWearingAHat commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+        ],
+        "item": ["sitrusberry"],
+        "nature": "impish",
+        "evs": [244, 4, 116, 0, 124, 20],
```
This 20 in speed isn't doing anything other than outspeeding other ting-lus since hippowdon doesn't exist here. Move at least most of it out to other stats. Same goes for the set below.

ACakeWearingAHat reviewed (changes requested):
These are a massive, massive amount of changes, but they're all for consistency, code standards, and clarity. Let me know when you're done looking and talking over these. Livid will review the code portion later.

livid-washed commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+          ["ceaselessedge"],
+          ["razorshell", "aquacutter"],
+          ["suckerpunch", "aquajet"],
+          ["sacredsword", "terablast"]
```
I don't think you want Tera Blast with Tera Type Water. Split this set?

livid-washed commented on data/bss-factory-sets.json:
```diff
@@ -0,0 +1,5689 @@
+          ["earthquake", "endeavor"],
+          ["ironhead"],
+          ["stealthrock", "rapidspin", "substitute"],
+          ["knockoff", "icespinner", "terablast"]
```
I don't think you want Tera Blast with Tera Type Ground.

LegoFigure11 commented:
Addressed all except the Roaring Moon set that I'm waiting to hear back on still

Sets were expanded/shrunk automatically by my linter based on line length, I've changed them all manually but it's gonna make maintaining this file a PITA. Is it really necessary? I would argue that it's more readable if they're all expanded anyway, but that's not my call to make.

ACakeWearingAHat reviewed (approved):
Sets are approved. Code review coming soon. Apologies for the wait.

livid-washed commented on data/random-teams.ts:
```diff
@@ -2448,6 +2448,293 @@ export class RandomTeams {
+			let hasRequiredMove = false;
+
+			// limit to 2 dedicated tera users per team
+			if (curSet.wantsTera &&	teamData.wantsTeraCount && teamData.wantsTeraCount >= maxWantsTera)	continue;
```
You need to update `teamData.wantsTeraCount` when updating team data in `randomBSSFactoryTeam()`, otherwise this check won't do anything.

LegoFigure11 commented:
Made all requested changes @livid-washed

livid-washed commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,299 @@ export class RandomTeams {
+
+		const maxWantsTera = 2;
+
+		const requiredMoves: { [k: string]: number } = {};
```
Since this is empty/unused, you should remove any code relating to `requiredMoves`, including `priorityPool`.

LegoFigure11 replied:
Great, thanks! It's outside the scope of this PR but `requiredMoves` and `priorityPool` also exist in Gen 8 (and I think Gen 7?) BSS Factory code too and are likewise unused there, so this change should probably be made in that code too.

livid-washed commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,299 @@ export class RandomTeams {
+			evs: {hp: 0, atk: 0, def: 0, spa: 0, spd: 0, spe: 0, ...setData.set.evs},
+			ivs: {hp: 31, atk: 31,	def: 31, spa: 31, spd: 31, spe: 31,	...setData.set.ivs},
+			nature: setData.set.nature || "Serious",
+			moves,
```
```suggestion
			moves,
			wantsTera: setData.set.wantsTera,
```

livid-washed reviewed (approved):
Code looks fine to me. Approved.

LegoFigure11 commented:
I do believe that's everything! Ready for merge, or further review if anyone else wants to take a look.

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+	randomBSSFactorySets: AnyObject = require("./bss-factory-sets.json");
+
+	randomBSSFactorySet(
+		species: Species,
+		teamData: RandomTeamsTypes.FactoryTeamDetails
```
```suggestion
		species: Species, teamData: RandomTeamsTypes.FactoryTeamDetails
```

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+		const id = toID(species.name);
+		const setList = this.randomBSSFactorySets[id].sets;
+
+		const movesMax: { [k: string]: number } = {
```
```suggestion
		const movesMax: {[k: string]: number} = {
```

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+			trickroom: 1,
+			auroraveil: 1,
+		};
+		const weatherAbilities = [
+			"drizzle",
+			"drought",
+			"snowwarning",
+			"sandstream",
+		];
```
```suggestion
		const weatherAbilities = ['drizzle', 'drought', 'snowwarning', 'sandstream'];
```

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+
+		// Build a pool of eligible sets, given the team partners
+		// Also keep track of sets with moves the team requires
+		const effectivePool: {set: AnyObject, moveVariants?: number[], itemVariants?: number, abilityVariants?: number}[] = [];
```
I don't like AnyObject here, please work on a type definition

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+			let reject = false;
+
+			// limit to 2 dedicated tera users per team
+			if (curSet.wantsTera &&	teamData.wantsTeraCount && teamData.wantsTeraCount >= maxWantsTera)	continue;
```
```suggestion
			if (curSet.wantsTera && teamData.wantsTeraCount && teamData.wantsTeraCount >= maxWantsTera) {
				continue;
			}
```

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+			// limit to 2 dedicated tera users per team
+			if (curSet.wantsTera &&	teamData.wantsTeraCount && teamData.wantsTeraCount >= maxWantsTera)	continue;
+
+			if (teamData.weather && weatherAbilities.includes(curSet.ability)) { continue; } // reject 2+ weather setters
```
```suggestion
			// reject 2+ weather setters
			if (teamData.weather && weatherAbilities.includes(curSet.ability)) {
				continue;
			}
```

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+
+		const setDataAbility = this.sampleIfArray(setData.set.ability);
+		return {
+			name: setData.set.nickname || setData.set.name || species.baseSpecies,
```
```suggestion
			name: setData.set.name || setData.set.species || species.baseSpecies,
```
Change this in the JSON as well please

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+			teraType: this.sampleIfArray(setData.set.teratype),
+			gender:	setData.set.gender || species.gender ||	(this.randomChance(1, 2) ? "M" : "F"),
+			item: this.sampleIfArray(setData.set.item) || "",
+			ability: setDataAbility || species.abilities["0"],
```
I feel like it would make more sense to make the ability go to "No Ability" in the event that setDataAbility doesn't exist, it would make it much easier to spot out errors

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+			ability: setDataAbility || species.abilities["0"],
+			shiny: typeof setData.set.shiny === "undefined" ? this.randomChance(1, 1024) : setData.set.shiny,
+			level: setData.set.level || 50,
+			happiness: typeof setData.set.happiness === "undefined" ? 255 :	setData.set.happiness,
```
is happiness even required for any comp aspect this gen? just set it to 255 always

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+			item: this.sampleIfArray(setData.set.item) || "",
+			ability: setDataAbility || species.abilities["0"],
+			shiny: typeof setData.set.shiny === "undefined" ? this.randomChance(1, 1024) : setData.set.shiny,
+			level: setData.set.level || 50,
```
don't levels get force adjusted to 50?

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+	}
+
+
+	randomBSSFactoryTeam(
+		side: PlayerOptions,
+		depth = 0
+	): RandomTeamsTypes.RandomFactorySet[] {
```
```suggestion
	randomBSSFactoryTeam(side: PlayerOptions, depth = 0): RandomTeamsTypes.RandomFactorySet[] {
```

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+			weaknesses: {},
+			resistances: {},
+		};
+		const weatherAbilitiesSet: { [k: string]: string } = {
```
```suggestion
		const weatherAbilitiesSet: {[k: string]: string} = {
```

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+			snowwarning: "hail",
+			sandstream: "sandstorm",
+		};
+		const resistanceAbilities: { [k: string]: string[] } = {
```
```suggestion
		const resistanceAbilities: {[k: string]: string[]} = {
```

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+		const shuffledSpecies = [];
+		for (const speciesName of pokemonPool) {
+			const sortObject = {
+				speciesName: speciesName,
```
```suggestion
				speciesName,
```

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+
+		while (shuffledSpecies.length && pokemon.length < this.maxTeamSize) {
+			// repeated popping from weighted shuffle is equivalent to repeated weighted sampling without replacement
+			const specie = shuffledSpecies.pop()!.speciesName;
+			const species = this.dex.species.get(specie);
```
```suggestion
			const species = this.dex.species.get(shuffledSpecies.pop()!.speciesName);
```

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+		return {
+			name: setData.set.nickname || setData.set.name || species.baseSpecies,
+			species: setData.set.species,
+			teraType: this.sampleIfArray(setData.set.teratype),
```
```suggestion
			teraType: this.sampleIfArray(setData.set.teraType),
```

KrisXV commented on data/random-teams.ts:
```diff
@@ -2448,6 +2449,292 @@ export class RandomTeams {
+			level: setData.set.level || 50,
+			happiness: typeof setData.set.happiness === "undefined" ? 255 :	setData.set.happiness,
+			evs: {hp: 0, atk: 0, def: 0, spa: 0, spd: 0, spe: 0, ...setData.set.evs},
+			ivs: {hp: 31, atk: 31,	def: 31, spa: 31, spd: 31, spe: 31, ...setData.set.ivs},
```
Won't this not work? IVs are an array in the JSON file

DaWoblefet commented:
Can you also update the /battlefactory help command to indicate support for the latest generation?

LegoFigure11 commented:
Sure! I believe Mathy is working on updating the command.

ACakeWearingAHat commented:
(moving to spotlight section temporarily was approved by rands leadership)

LegoFigure11 commented:
Sorry for review requests, I botched a rebase

KrisXV commented on config/formats.ts:
```diff
@@ -2491,6 +2491,17 @@ export const Formats: FormatList = [
+		mod: 'gen9',
+		team: 'randomBSSFactory',
+		ruleset: ['Flat Rules', 'VGC Timer'],
+  },
+  {
```
```suggestion
	},
	{
```
